### PR TITLE
Add s3 storage artifact route and ui integration of it

### PR DESCRIPTION
**Pull request by Gkrumbach07**
closes: https://issues.redhat.com/browse/RHOAIENG-7081

## Description

- created a new /api/storage GET route that takes bucket and namespace along with the path as a query param and gives back what is in that s3 bucket.
- using that in compare runs markdown compare and in artifacts list uri link
- added preview and peek file size warnings

Screenshot 2024-06-05 at 4 32 42 PM
Screenshot 2024-06-05 at 4 12 18 PM
Screenshot 2024-06-05 at 4 11 55 PM
Screenshot 2024-06-05 at 4 05 31 PM

## How Has This Been Tested?

- go to artifacts list page and navigate to pipelines with output artifacts
   (reach out to @Gkrumbach07 i you need a test cluster for this PR)

## Test Impact

no tests add bc this is a backend route, will add util tests

## Request review criteria:

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 

- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

cc @yannnz

### CODE REVIEW

**Comment by Gkrumbach07** [comment]
keeping this as WIP until security review is done. I will also add tests in the meantime

File: backend/src/routes/api/storage/index.ts
Comment (andrewballantyne) [inline backend/src/routes/api/storage/index.ts:10]: Key shouldn't be here, right? key is in the query...

File: backend/src/routes/api/storage/index.ts
Comment (andrewballantyne) [inline backend/src/routes/api/storage/index.ts:5]: Technically the Request & Reply are generics

File: backend/src/routes/api/storage/index.ts
Comment (Gkrumbach07) [reply backend/src/routes/api/storage/index.ts]: fixed

File: backend/src/routes/api/storage/index.ts
Comment (Gkrumbach07) [reply backend/src/routes/api/storage/index.ts]: fixed, i think

**Comment by Gkrumbach07** [comment]
@andrewballantyne I took this out of WIP, the safety check for the storage status is used, and the feature flag blocks the endpoint

File: frontend/src/concepts/pipelines/content/artifacts/ArtifactUriLink.tsx
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/artifacts/ArtifactUriLink.tsx:34]: Not sure what this comment is telling that the code immediately under it doesn't show almost directly.

File: backend/src/routes/api/storage/storageUtils.ts
Comment (andrewballantyne) [inline backend/src/routes/api/storage/storageUtils.ts:14]: This is problematic -- this is done with the service account. We need to do this with the user's token.

Otherwise you could provide a namespace you do not have access to and get back a response.

See the `utils/proxy.ts` for an example.

File: backend/src/routes/api/storage/storageUtils.ts
Comment (andrewballantyne) [inline backend/src/routes/api/storage/storageUtils.ts:54]: This should also be done with the user's token

File: backend/src/routes/api/storage/index.ts
Comment (andrewballantyne) [inline backend/src/routes/api/storage/index.ts:7]: It's a stupid config -- sadly, this won't work.

disableS3Endpoint equals:
* `true` (flag off) => !true => false ✅ 
* `false` (flag on) => !false => true ✅ 
* `undefined` (flag not set) => !undefined => true ❌  

Explicitly check to see if it's `=== false`

**Review by andrewballantyne — CHANGES_REQUESTED** [review]
If you would like to talk about permission flows more, let me know

File: frontend/src/concepts/pipelines/content/artifacts/ArtifactUriLink.tsx
Comment (andrewballantyne) [inline frontend/src/concepts/pipelines/content/artifacts/ArtifactUriLink.tsx:34]: Codecov says you don't have a test for this -- can you make sure you have a mocked cypress test for this?

File: backend/src/routes/api/storage/index.ts
Comment (andrewballantyne) [inline backend/src/routes/api/storage/index.ts:53]: My leading concern -- which I'm not sure we need to address it here -- but should before we release it. 

If the resource you're pulling is so large the data crashes our pod, it's not going to be awesome. You're also holding up the connection until you're done -- which may time out.

We should probably find a way to peek only the first "bunch" of characters. Are we responsible for the "full download / view in a new tab"?

Comment (andrewballantyne) [reply backend/src/routes/api/storage/index.ts:53]: It would appear we are on the hook for the new tab too... if my scanning of the code was accurate

File: frontend/src/services/storageService.ts
Comment (andrewballantyne) [inline frontend/src/services/storageService.ts:1]: fwiw, you don't need to use the services folder -- I went back and forth on this, but I think I need to deprecate this folder entirely.

File: backend/src/routes/api/storage/index.ts
Comment (Gkrumbach07) [reply backend/src/routes/api/storage/index.ts:53]: As far as i could tell, the KF ui will attempt to pull all in the new tab unless told to preview

File: backend/src/routes/api/storage/index.ts
Comment (Gkrumbach07) [inline backend/src/routes/api/storage/index.ts:19]: switch condition

File: backend/src/routes/api/storage/index.ts
Comment (Gkrumbach07) [inline backend/src/routes/api/storage/index.ts:14]: take bucket out

File: backend/src/routes/api/storage/index.ts
Comment (andrewballantyne) [inline backend/src/routes/api/storage/index.ts:10]: Remove bucket, get from DSPA -- maybe make `createMinioClient` into a `setupMinioClient` function that returns the bucket and the client... then use them in the `getObjectStream`.

Comment (Gkrumbach07) [inline backend/src/routes/api/storage/index.ts:10]: create new endpoint for getting size

File: backend/src/routes/api/storage/index.ts
Comment (andrewballantyne) [inline backend/src/routes/api/storage/index.ts:27]: Can you get the stat information on the file before downloading it so we know how large it is... maybe make that another endpoint and allow this one as a text-stream endpoint.

File: backend/src/routes/api/storage/index.ts
Comment (andrewballantyne) [inline backend/src/routes/api/storage/index.ts:38]: Use Peek to avoid getting more than 100 megabtypes...

File: frontend/src/concepts/pipelines/content/artifacts/ArtifactUriLink.tsx
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/content/artifacts/ArtifactUriLink.tsx]: there is a jest test for it. it is hard to create a cypress test becuase this data is from mlmd

File: backend/src/utils/constants.ts
Comment (andrewballantyne) [inline backend/src/utils/constants.ts:65]: default needs to be `true`

File: backend/src/routes/api/storage/index.ts
Comment (andrewballantyne) [inline backend/src/routes/api/storage/index.ts:64]: We need to make sure peek doesn't exceed our limit -- it doesn't matter what the UI says

**Review by andrewballantyne — APPROVED** [review]
Okay, let's go with this implementation. We'll be looking at replacing this logic with more structured backend soon, so this is temporary. 

We need loading spinners for the artifact peek loading & we definitely need to consider why the Artifact page does not have the peek -- it's awkward that the direct source (the artifact page itself) of the data doesn't have it but the embedded sources (on the runs page) do... 

/hold 

Not sure if you're addressing the loading spinners now or later, unhold it if you'll follow up in another PR.

**Review by andrewballantyne — APPROVED** [review]
lgtm

/unhold

**Pull request merged**
